powerauth-push-server/src/main/java/io/getlime/push/service/PushMessageSenderService.java

```diff
@@ -278,6 +293,29 @@ private void updateStatusAndPersist(PushMessageEntity pushMessageObject, PushMes
         }
     }
 
+    private ApnsEnvironment resolveApnsEnvironment(final String environmentDevice, final String environmentAppConfig) {
+        if (environmentDevice != null) {
```

Comment:
Just as a code concept for explaining the resolution logic - the switching between an enum and a String is hard to follow. We should not use the enum on the Service layer at all IMHO. @banterCZ 

// resolve environment from the application configuration
String environment = environmentAppConfig;
// resolve environment from the server configuration if not set on application level
if (environment == null) {
 environment = configuration.isApnsUseDevelopment();
}
// now the environment should be already set - otherwise we have invalid configuration, but let's skip it.

// the environment is set on the device level
if (environmentDevice != null) {
// the device is set for production or the server / app is configured for development
 if (environmentDevice == ApnsEnvironment.PRODUCTION || environment == ApnsEnvironment.DEVELOPMENT) {
    // we can do the validation in this way
    environment = environmentDevice;
  } 
}

return environment;

Reply:
Ok, I will use this logic. Based on test results it is identical to original code, but it seems more readable.